# #  WEBVTT

# 00:00:06.210 --> 00:00:11.010
# Hello and welcome to the introduction to this lecture and this lecture we're going to briefly discuss

# 00:00:11.010 --> 00:00:16.140
# what pandas is how to get it installed onto your computer and what to expect throughout this section

# 00:00:16.140 --> 00:00:17.940
# of the course.

# 00:00:18.150 --> 00:00:23.810
# Pandas is an open source library that's built on top of empire which is why we learned but PI first

# 00:00:23.810 --> 00:00:24.270
# .

# 00:00:24.510 --> 00:00:28.490
# Panthers allows for fast analysis and data cleaning and preparation.

# 00:00:28.500 --> 00:00:34.350
# A lot of people like to call pandas pythons version of Excel or if you're familiar with our Python's

# 00:00:34.350 --> 00:00:40.670
# version of our data frames fun and pandas really excels in performance and productivity for the user

# 00:00:40.680 --> 00:00:40.920
# .

# 00:00:41.150 --> 00:00:45.960
# And this also has built in data visualization features which we'll learn about once he gets a python

# 00:00:45.960 --> 00:00:52.670
# for data visualization Pandurs can also work with data from a wide variety of sources.

# 00:00:52.880 --> 00:00:58.080
# You'll need to install Pandurs by going to your command line or terminal and then use either Konda install

# 00:00:58.080 --> 00:00:58.680
# pandas.

# 00:00:58.710 --> 00:01:03.750
# If you have the Anaconda distribution of Python or if you installed Python through a different method

# 00:01:04.050 --> 00:01:06.680
# you can also use Pipp install pandas.

# 00:01:06.720 --> 00:01:08.720
# Either one should work for you.

# 00:01:09.810 --> 00:01:14.910
# As far as what to expect throughout this section of course we'll first discuss series and how they interact

# 00:01:14.910 --> 00:01:15.890
# with pandas.

# 00:01:16.050 --> 00:01:19.430
# Then we'll build on top of that to go to data frames with pandas.

# 00:01:19.440 --> 00:01:24.720
# Then we'll discuss some particular topics such as work of missing data how to use GROUP BY of pandas

# 00:01:25.110 --> 00:01:31.140
# merging joining in cannonading data frames of pandas various pandas operations and then how to get data

# 00:01:31.230 --> 00:01:38.340
# in and out of pandas for a wide variety of forms such as CSB files Excel files sequel and more.

# 00:01:38.340 --> 00:01:41.300
# All right let's go ahead and get started with pandas.

# 00:01:41.460 --> 00:01:43.540
# Thanks and I'll see you at the next lecture.

# OUTLINES:
# (1) SERIES
# (2) DATAFRAMES
# (3)MISSING DATA
# (4) GROUP BY
# (5) MERGING, JOINING, AND CONCATENATNG
# (6) OPERATIONS
# (7) DATA INPUT AND OUTPUTS


# (1) series in padans operations: this is very similar to Numpy arrays, the only diff. is that series csn have access on labels, ie can be indexe by label

In [21]:
import numpy as np

In [22]:
import pandas as pd

In [23]:
labels = ['a', 'b', 'c']
my_data = [10,20,30]
arr = np.array(my_data)
d = {'a':10, 'b':20, 'c':30}


# creating Series operation

In [24]:
pd.Series(data = my_data)

0    10
1    20
2    30
dtype: int64

# trying to specify what you want the index to be (key)

In [25]:
pd.Series(data = my_data, index=labels)

a    10
b    20
c    30
dtype: int64

In [26]:
# short-cut

pd.Series(my_data, labels)

a    10
b    20
c    30
dtype: int64

In [27]:
# more on Series()
pd.Series(d)

a    10
b    20
c    30
dtype: int64

# Series can actually hold many any type of objects, been it built-in functions etc



In [28]:
pd.Series(data=[sum,print,len])

0      <built-in function sum>
1    <built-in function print>
2      <built-in function len>
dtype: object

In [29]:
ser1 = pd.Series([1,2,3,4],['USA','Germany', 'Canada', 'Japan'])

In [30]:
ser1

USA        1
Germany    2
Canada     3
Japan      4
dtype: int64

In [31]:
ser2 = pd.Series([1,2,3,4],['USA','Germany', 'Canada', 'China'])

In [32]:
ser2

USA        1
Germany    2
Canada     3
China      4
dtype: int64

# trying to grab information on the series is actually going to be almost thesame tin in python, the main duty of Series is to hold label indexes

In [33]:
ser1['USA']

1

In [34]:
ser3 = pd.Series(data=labels)

In [35]:
ser3[0]

'a'

# adding two series: why the NaN, and floats

In [36]:
ser1 + ser2

Canada     6.0
China      NaN
Germany    4.0
Japan      NaN
USA        2.0
dtype: float64

# when you're working with series under any mathematical computations it's defnatly going to be converted in to float

# WORKING ON DATA FRAMES:
# DataFrame is all about bunch of series that share thesame index
# making sure you import the above, numpy and pandas. this is also the main tool while with Data science

In [37]:
  from numpy.random import randn

# seed helps us to get thesame random numbers


In [38]:
np.random.seed(101)

In [39]:
df = pd.DataFrame(randn(5,4),['A','B','C','D','E'],['W','X','Y','Z'])

In [40]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

# selecting or grabbing some columns in the table above

In [41]:
# df['W']

In [42]:
# type(df['W'])

In [43]:
# type(df)

# another of grabbing an item or column in a DataFrame is (CQL)

In [44]:
# its not advisible to use this method df.X , rather use the other one
# df['X']

# grabbing double 

In [45]:
# df[['W','Z']]

# creating a new column, if it's alredy existing ie the df

In [46]:
df['new'] = df['W'] + df['Y']

In [47]:
df

W         X         Y         Z       new
A  2.706850  0.628133  0.907969  0.503826  3.614819
B  0.651118 -0.319318 -0.848077  0.605965 -0.196959
C -2.018168  0.740122  0.528813 -0.589001 -1.489355
D  0.188695 -0.758872 -0.933237  0.955057 -0.744542
E  0.190794  1.978757  2.605967  0.683509  2.796762

# for removing columns

In [48]:
df.drop('new',axis=1)

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [49]:
df

W         X         Y         Z       new
A  2.706850  0.628133  0.907969  0.503826  3.614819
B  0.651118 -0.319318 -0.848077  0.605965 -0.196959
C -2.018168  0.740122  0.528813 -0.589001 -1.489355
D  0.188695 -0.758872 -0.933237  0.955057 -0.744542
E  0.190794  1.978757  2.605967  0.683509  2.796762

# now you can see that this particular drop methos is not deleteting or removing the paricular element you want. here's the solution

In [50]:
df.drop('new',axis=1,inplace=True)

In [51]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

# inplace=True, this function helps us to still maintain some data, without loosing 
# them, Note without using the function, the data you're trying to drop will still be there once you call on the variable name
# the axis is to specify that you're dropping a column
# by default the rows axis=0, while the column axis=1


# Dropping a row: by default the axis=0, so you sont don't need to add it once you're working with rows



In [38]:
df.drop('E')

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057

In [39]:
# df.drop('E', inplace=True)

In [40]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

In [41]:
# df.shape

# lets talk about selecting rows: there are about two ways in selecting ROWS in padas and they're (loc,as location), and (iloc)

# this locates the mentioned row, accoding to the name
# this is known as location of label index


In [42]:
df.loc['A']

W    2.706850
X    0.628133
Y    0.907969
Z    0.503826
Name: A, dtype: float64

# iloc, this locates the index according the index specified
# this is known as location of numerical index


In [43]:
df.iloc[2]

W   -2.018168
X    0.740122
Y    0.528813
Z   -0.589001
Name: C, dtype: float64

# selecting of subset of rows and columns

In [44]:
df.loc['B','Y']

-0.8480769834036315

In [45]:
df


W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

# selecting list subset of rows and column, such as A and B rows with subset of Y

In [46]:
df.loc[['A','B'],['W','Y']]

W         Y
A  2.706850  0.907969
B  0.651118 -0.848077

In [47]:
my_tables = np.arange(25)

In [48]:
my_tables

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24])

In [49]:
my_tables.reshape(5,5)

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

In [50]:
condition = my_tables > 10

In [51]:
condition 

array([False, False, False, False, False, False, False, False, False,
       False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True])

In [52]:
my_tables[condition]

array([11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24])

In [53]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

# conditional selection, using bracket notation

In [54]:
booldf = df > 0

In [55]:
booldf

W      X      Y      Z
A   True   True   True   True
B   True  False  False   True
C  False   True   True  False
D   True  False  False   True
E   True   True   True   True

In [56]:
df[booldf]

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118       NaN       NaN  0.605965
C       NaN  0.740122  0.528813       NaN
D  0.188695       NaN       NaN  0.955057
E  0.190794  1.978757  2.605967  0.683509

# now lets try to only return the ones that are True not NaN

In [57]:
resultdf = df[df['W']>0]

In [58]:
resultdf['X']

A    0.628133
B   -0.319318
D   -0.758872
E    1.978757
Name: X, dtype: float64

# trying to return all the rows that's greater than 0

In [59]:
df[df['W']>0][['Y', 'X']]

Y         X
A  0.907969  0.628133
B -0.848077 -0.319318
D -0.933237 -0.758872
E  2.605967  1.978757

# breaking it down

In [60]:
boolser = df['W']>0 #print it 1st
result = df[boolser]
mycols = ['Y', 'X']
result[mycols]

Y         X
A  0.907969  0.628133
B -0.848077 -0.319318
D -0.933237 -0.758872
E  2.605967  1.978757

# the concept of using multiple conditional selectors, ie trying to get double answers in a condition, let say returning tables where W>0 and Y>1

In [61]:
df[(df['W']>0) & (df['Y']>1)] # note do not use the and while working with DataFrame

W         X         Y         Z
E  0.190794  1.978757  2.605967  0.683509

# now lets talk about index in a more details, reseting index to smtin else, lets say to be numerical format

In [62]:
df.reset_index()

index         W         X         Y         Z
0     A  2.706850  0.628133  0.907969  0.503826
1     B  0.651118 -0.319318 -0.848077  0.605965
2     C -2.018168  0.740122  0.528813 -0.589001
3     D  0.188695 -0.758872 -0.933237  0.955057
4     E  0.190794  1.978757  2.605967  0.683509

## note the rest_index does not actually affect the original index.

In [63]:
df

W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509

# for this to actually occur inplace:

In [64]:
df.reset_index(inplace=True)

In [65]:
df

index         W         X         Y         Z
0     A  2.706850  0.628133  0.907969  0.503826
1     B  0.651118 -0.319318 -0.848077  0.605965
2     C -2.018168  0.740122  0.528813 -0.589001
3     D  0.188695 -0.758872 -0.933237  0.955057
4     E  0.190794  1.978757  2.605967  0.683509

In [66]:
# df

### more examples on indexes are:

In [67]:
newind = 'CA NY WY OR CO'.split()


In [68]:
newind

['CA', 'NY', 'WY', 'OR', 'CO']

In [69]:
df['States'] = newind

In [70]:
df

index         W         X         Y         Z States
0     A  2.706850  0.628133  0.907969  0.503826     CA
1     B  0.651118 -0.319318 -0.848077  0.605965     NY
2     C -2.018168  0.740122  0.528813 -0.589001     WY
3     D  0.188695 -0.758872 -0.933237  0.955057     OR
4     E  0.190794  1.978757  2.605967  0.683509     CO

### now lets use the set method to set the States to be index of the df. note this will overide ur previous index unless you retain the former

In [71]:
df.set_index('States',inplace=True)

In [72]:
df

index         W         X         Y         Z
States                                              
CA         A  2.706850  0.628133  0.907969  0.503826
NY         B  0.651118 -0.319318 -0.848077  0.605965
WY         C -2.018168  0.740122  0.528813 -0.589001
OR         D  0.188695 -0.758872 -0.933237  0.955057
CO         E  0.190794  1.978757  2.605967  0.683509

# Pandas - DataFrame part 3. MultiIndex. let's talk about (index Levels)

In [73]:
outside = ['G1', 'G1','G1', 'G2', 'G2', 'G2']
inside = [1,2,3,1,2,3]
hier_index = list(zip(outside, inside))
hier_index = pd.MultiIndex.from_tuples(hier_index)
# u can print them individually
# the zip() function made them to look like tupples()

In [74]:
my_df = pd.DataFrame(randn(6,2),hier_index,['A','B'])

# the index is the hier_index we made, while the columns are the 'A', and 'B'

In [75]:
my_df

A         B
G1 1  0.302665  1.693723
   2 -1.706086 -1.159119
   3 -0.134841  0.390528
G2 1  0.166905  0.184502
   2  0.807706  0.072960
   3  0.638787  0.329646

# the above table shows how a MultiIndex looks like, otherwise known as index hierachy

# let show how we can call data from index hierachy

In [76]:
my_df.loc['G1']

A         B
1  0.302665  1.693723
2 -1.706086 -1.159119
3 -0.134841  0.390528

# given them names

In [77]:
my_df.index.names

FrozenList([None, None])

# the above result shows that it has no name yet

In [78]:
my_df.index.names = ['Groups', 'Num']

In [79]:
my_df

A         B
Groups Num                    
G1     1    0.302665  1.693723
       2   -1.706086 -1.159119
       3   -0.134841  0.390528
G2     1    0.166905  0.184502
       2    0.807706  0.072960
       3    0.638787  0.329646

# lets talk about how we can grab information from the data once again. example lets grab G2 in B 0.072960

In [80]:
my_df.loc['G2'].loc[2]['B']

0.07295967531703869

# cross section using a function called xs. and this is use when you're working with multi section data

In [81]:
my_df.xs

<bound method NDFrame.xs of                    A         B
Groups Num                    
G1     1    0.302665  1.693723
       2   -1.706086 -1.159119
       3   -0.134841  0.390528
G2     1    0.166905  0.184502
       2    0.807706  0.072960
       3    0.638787  0.329646>

# now lets try using the xs to grab from the data anythig thta has 1 and its values

In [82]:
my_df.xs(1,level='Num')

A         B
Groups                    
G1      0.302665  1.693723
G2      0.166905  0.184502

#  MISSING DATA LECTURE IN PANDAS

# allots of times when you use pandas to read data,if there's a missing point, what pandas is going to do is to filled in the missing points with Non or any n value. now lets see how we can use a methods like drop or fill to drop or fill in those missing vslues

In [83]:
my_dic = {'A':[1,2,np.nan], 'B':[5,np.nan,np.nan], 'C':[6,7,8]}

In [84]:
dataFr = pd.DataFrame(my_dic)

In [85]:
dataFr

A    B  C
0  1.0  5.0  6
1  2.0  NaN  7
2  NaN  NaN  8

# using a drop nan method

In [86]:
dataFr.dropna()

A    B  C
0  1.0  5.0  6

# now lets drop entirely any row and column that has empty row

In [87]:
dataFr.dropna(axis=1)

C
0  6
1  7
2  8

# another way of dropping nan >1

In [88]:
dataFr.dropna(thresh=2)

A    B  C
0  1.0  5.0  6
1  2.0  NaN  7

# let's talk about the fill nan methods

In [89]:
dataFr.fillna(value='FILL VALUE')

A           B  C
0         1.0         5.0  6
1         2.0  FILL VALUE  7
2  FILL VALUE  FILL VALUE  8

## FIlling in the missing vslues with the mean() method

In [90]:
dataFr['A'].fillna(value=dataFr['A'].mean())

0    1.0
1    2.0
2    1.5
Name: A, dtype: float64

In [91]:
# there are allots of ways to filled in an empty space 

# GROUPBY  METHOD: This method allows us to group together rows based off of a column an aggregate funtion on them. the aggreagate fuction take in many values and return a single value as an output  

In [92]:
data = {'Company':['GOOG', 'GOOG', 'MFST', 'MFST', 'FB', 'FB'],
       'Person':['patterson', 'Chinedum', 'comfort', 'Promise', 'Patience', 'Blessing'],
        'Sales':[20000,120,356,231,900, 245]}

In [93]:
my_groupB = pd.DataFrame(data)

In [94]:
my_groupB

Company     Person  Sales
0    GOOG  patterson  20000
1    GOOG   Chinedum    120
2    MFST    comfort    356
3    MFST    Promise    231
4      FB   Patience    900
5      FB   Blessing    245

# lets groupby the company

In [95]:
by_com = my_groupB.groupby('Company')

In [96]:
by_com.mean()

Sales
Company         
FB         572.5
GOOG     10060.0
MFST       293.5

# note you can't do that in Person, bcos they're strings. lets get the sum() of everthing

In [97]:
by_com.sum()

Sales
Company       
FB        1145
GOOG     20120
MFST       587

# for there standard deviations

In [98]:
by_com.std()

Sales
Company              
FB         463.154942
GOOG     14057.282810
MFST        88.388348

# getting the sum of a particular company

In [99]:
by_com.sum().loc['FB']

Sales    1145
Name: FB, dtype: int64

# doing all this without using a variable, just a straight fawward

In [100]:
my_groupB.groupby('Company').sum().loc['GOOG']

Sales    20120
Name: GOOG, dtype: int64

# other useful aggregate operations that we might be using continually are:

# Count() function. this will return the Person, bcos they can be counted unlike other funtions

In [101]:
my_groupB.groupby('Company').count()

Person  Sales
Company               
FB            2      2
GOOG          2      2
MFST          2      2

In [102]:
my_groupB.groupby('Company').max()

Person  Sales
Company                  
FB        Patience    900
GOOG     patterson  20000
MFST       comfort    356

In [103]:
my_groupB.groupby('Company').min()

Person  Sales
Company                 
FB       Blessing    245
GOOG     Chinedum    120
MFST      Promise    231

#  the describe method: this returns a useful information about the data

In [104]:
my_groupB.groupby('Company').describe()

Sales                                                            \
        count     mean           std    min      25%      50%       75%   
Company                                                                   
FB        2.0    572.5    463.154942  245.0   408.75    572.5    736.25   
GOOG      2.0  10060.0  14057.282810  120.0  5090.00  10060.0  15030.00   
MFST      2.0    293.5     88.388348  231.0   262.25    293.5    324.75   

                  
             max  
Company           
FB         900.0  
GOOG     20000.0  
MFST       356.0

# if that didi't give you what u really want you can then transpose them

In [105]:
my_groupB.groupby('Company').describe().transpose()

Company              FB         GOOG        MFST
Sales count    2.000000      2.00000    2.000000
      mean   572.500000  10060.00000  293.500000
      std    463.154942  14057.28281   88.388348
      min    245.000000    120.00000  231.000000
      25%    408.750000   5090.00000  262.250000
      50%    572.500000  10060.00000  293.500000
      75%    736.250000  15030.00000  324.750000
      max    900.000000  20000.00000  356.000000

# you can actually get a particular description of a particular company

In [106]:
my_groupB.groupby('Company').describe().transpose()['FB']

Sales  count      2.000000
       mean     572.500000
       std      463.154942
       min      245.000000
       25%      408.750000
       50%      572.500000
       75%      736.250000
       max      900.000000
Name: FB, dtype: float64

# MERGING, JOINING, AND CONCATENATING

# We are going to learn how to combine variety of methods

In [107]:
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
        'B': ['B0', 'B1', 'B2', 'B3'],
        'C': ['C0', 'C1', 'C2', 'C3'],
        'D': ['D0', 'D1', 'D2', 'D3']},
        index = [0, 1, 2, 3])

In [108]:
df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
        'B': ['B4', 'B5', 'B6', 'B7'],
        'C': ['C4', 'C5', 'C6', 'C7'],
        'D': ['D4', 'D5', 'D6', 'D7']},
        index = [4, 5, 6, 7])

In [109]:
df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
        'B': ['B8', 'B9', 'B10', 'B11'],
        'C': ['C8', 'C9', 'C10', 'C11'],
        'D': ['D8', 'D9', 'D10', 'D11']},
        index = [8, 9, 10, 11])

In [110]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [111]:
df2

A   B   C   D
4  A4  B4  C4  D4
5  A5  B5  C5  D5
6  A6  B6  C6  D6
7  A7  B7  C7  D7

In [112]:
df3

A    B    C    D
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

# concatenation: thi is basically glues together DataFrames, keep in mind that dimmensions should match along the axis you are concatinating on. You can use pd.concat() and pass in list of DataFrame to cancatenate together

In [113]:
pd.concat([df1,df2,df3])

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

# concatenation based on axis=1. 
* before you cantenate make sure you have all the values available.
* else, it will be filled NaN to the empty spaces in the axis 

In [114]:
pd.concat([df1,df2,df3],axis=1)

A    B    C    D    A    B    C    D    A    B    C    D
0    A0   B0   C0   D0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1    A1   B1   C1   D1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2    A2   B2   C2   D2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3    A3   B3   C3   D3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
4   NaN  NaN  NaN  NaN   A4   B4   C4   D4  NaN  NaN  NaN  NaN
5   NaN  NaN  NaN  NaN   A5   B5   C5   D5  NaN  NaN  NaN  NaN
6   NaN  NaN  NaN  NaN   A6   B6   C6   D6  NaN  NaN  NaN  NaN
7   NaN  NaN  NaN  NaN   A7   B7   C7   D7  NaN  NaN  NaN  NaN
8   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A8   B8   C8   D8
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A9   B9   C9   D9
10  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A10  B10  C10  D10
11  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A11  B11  C11  D11

# More examples on Merging Fuction

In [115]:
left = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3'],
        'A': ['A0', 'A1', 'A2', 'A3'],
        'B': ['B0', 'B1', 'B2', 'B3']})

right = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3'],
        'C': ['C0', 'C1', 'C2', 'C3'],
        'D': ['D0', 'D1', 'D2', 'D3']})
        

In [116]:
left

key   A   B
0  K0  A0  B0
1  K1  A1  B1
2  K2  A2  B2
3  K3  A3  B3

In [117]:
right

key   C   D
0  K0  C0  D0
1  K1  C1  D1
2  K2  C2  D2
3  K3  C3  D3

# Merging Fuction: this allows DataFrames together using logic as merging SQL Table together according to there key columns. such as:

In [118]:
pd.merge(left,right,how='inner',on='key')

key   A   B   C   D
0  K0  A0  B0  C0  D0
1  K1  A1  B1  C1  D1
2  K2  A2  B2  C2  D2
3  K3  A3  B3  C3  D3

# or to show a more complicated example

In [119]:
left = pd.DataFrame({'key1': ['K0', 'K0', 'K1', 'K2'],
                     'key2': ['K0', 'K1', 'K0', 'K1'],
                      'A': ['A0', 'A1', 'A2', 'A3'],
                      'B': ['B0', 'B1', 'B2', 'B3']})

right = pd.DataFrame({'key1': ['K0', 'K1', 'K1', 'K2'],
                      'key2': ['K0', 'K0', 'K0', 'K0'],
                       'C': ['C0', 'C1', 'C2', 'C3'],
                       'D': ['D0', 'D1', 'D2', 'B3']})

In [120]:
pd.merge(left,right, on=['key1', 'key2'])

key1 key2   A   B   C   D
0   K0   K0  A0  B0  C0  D0
1   K1   K0  A2  B2  C1  D1
2   K1   K0  A2  B2  C2  D2

# adding how and on

In [121]:
pd.merge(left,right, how = 'outer', on=['key1', 'key2'])

key1 key2    A    B    C    D
0   K0   K0   A0   B0   C0   D0
1   K0   K1   A1   B1  NaN  NaN
2   K1   K0   A2   B2   C1   D1
3   K1   K0   A2   B2   C2   D2
4   K2   K1   A3   B3  NaN  NaN
5   K2   K0  NaN  NaN   C3   B3

In [122]:
pd.merge(left,right, how = 'right', on=['key1', 'key2'])

key1 key2    A    B   C   D
0   K0   K0   A0   B0  C0  D0
1   K1   K0   A2   B2  C1  D1
2   K1   K0   A2   B2  C2  D2
3   K2   K0  NaN  NaN  C3  B3

# don't too worry about those how, inner, right parameters, unless you're good in SQL,although the add missing numbers, let's just be using the previous.

# Joining: this is a convinient method for combining the columns of two potentially differently-indexed DataFrame into a ingle result

In [123]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2'],
                     'B': ['B0', 'B1', 'B2']},
                      index=['K0', 'K1', 'K3'])


right = pd.DataFrame({'C': ['C0', 'C2', 'C3'],
                     'D': ['D0', 'D2', 'D3']},
                      index=['K0', 'K1', 'K3'])
        
        

In [124]:
left.join(right)

A   B   C   D
K0  A0  B0  C0  D0
K1  A1  B1  C2  D2
K3  A2  B2  C3  D3

In [125]:
left.join(right, how='outer')

A   B   C   D
K0  A0  B0  C0  D0
K1  A1  B1  C2  D2
K3  A2  B2  C3  D3

# OPERATIONS IN PANDAS:
* in this lecture we'll be covering some Pandas operations we've not been using for a while now

In [126]:
new_df = pd.DataFrame({'col1':[1,2,3,4],
                       'col2':[444,555,666,444],
                       'col3':['abc', 'def', 'ghi', 'xyz']})

new_df.head()

col1  col2 col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

# fining unique values in a DataFrame. and there about three ways of doing that and they're
* unique(): this returns all the unique values in the table as Numpy arrays form.
* and can as well return the lenght with nunique() methods
* another method is the value_counts(), which returns the data in form of table  

In [127]:
new_df['col2'].unique() # for the lenght .nunique()

array([444, 555, 666], dtype=int64)

# value_counts() method

In [128]:
new_df['col2'].value_counts()

444    2
666    1
555    1
Name: col2, dtype: int64

# Selecting Data: more and powerful ways of using conditional selecting data

In [129]:
new_df

col1  col2 col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

In [130]:
new_df[new_df['col1']>2]

col1  col2 col3
2     3   666  ghi
3     4   444  xyz

# combining conditions, you must wrap them with parenthese ()

In [131]:
new_df[(new_df['col1']>2) & (new_df['col2']==444)]

col1  col2 col3
3     4   444  xyz

# Applied method()

In [132]:
def times2(x):
    return x*2

In [133]:
new_df['col1'].sum()

10

In [134]:
# assumming we want to apply our function to it

In [135]:
new_df['col1'].apply(times2)

0    2
1    4
2    6
3    8
Name: col1, dtype: int64

# applying bult-in functions such as :

In [136]:
new_df['col3'].apply(len)

0    3
1    3
2    3
3    3
Name: col3, dtype: int64

# apply is a powerful function which we'll be using often. 
* and much powerful if applied with lambda function, which helps us not to write much code

In [137]:
new_df['col2'].apply(lambda X: X*2)

0     888
1    1110
2    1332
3     888
Name: col2, dtype: int64

# the above function is actually what we'll be using often on Panda expressions

# let's talk about remocing columns, like removing column one from new_df

In [138]:
new_df.drop('col1', axis=1) # you can apply your inplace=True

col2 col3
0   444  abc
1   555  def
2   666  ghi
3   444  xyz

In [139]:
new_df

col1  col2 col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

# checking the columns and the indexes

In [140]:
new_df.columns

Index(['col1', 'col2', 'col3'], dtype='object')

In [141]:
new_df.index

RangeIndex(start=0, stop=4, step=1)

# Sorting and ordering in Pandas.
* let's sort the the table, making sure the two 444 comes before 555

In [142]:
new_df

col1  col2 col3
0     1   444  abc
1     2   555  def
2     3   666  ghi
3     4   444  xyz

In [143]:
new_df.sort_values(by='col2')

col1  col2 col3
0     1   444  abc
3     4   444  xyz
1     2   555  def
2     3   666  ghi

# useful method to find a null value in Pandas

In [144]:
new_df.isnull()

col1   col2   col3
0  False  False  False
1  False  False  False
2  False  False  False
3  False  False  False

In [145]:
data = {'A':['foo', 'foo', 'foo', 'bar', 'bar', 'bar'],
         'B':['one', 'one', 'two', 'two', 'one', 'one'],
         'C':['X', 'Y', 'X', 'Y', 'X', 'Y'],
          'D':[1,3,2,5,4,1]}

new_df = pd.DataFrame(data) 

In [146]:
new_df

A    B  C  D
0  foo  one  X  1
1  foo  one  Y  3
2  foo  two  X  2
3  bar  two  Y  5
4  bar  one  X  4
5  bar  one  Y  1

# creating a pivot table

In [147]:
new_df.pivot_table(values='D', index=['A', 'B'],columns=['C'])

C          X    Y
A   B            
bar one  4.0  1.0
    two  NaN  5.0
foo one  1.0  3.0
    two  2.0  NaN

# DATA INPUT AND OUTPUT.
* Pandas as liabrary has the ability of reading and writing data and give the outpus.
### and there about four ways:
1. CSV files
2. Excel
3. HTML
4. SQL

# inother to wok with sql and html files, you'll have to install the following using conda or pip:

* install sqlalchemy
* install lxml
* install html5lib
* install beautifulSoup4

# read and write files in Pandas

# checking your location

In [148]:
pwd

'C:\\Users\\CHINEDUM\\Desktop\\PATTERSON-CODES\\Data-science'

# how to open and read CSV files

In [154]:
read = pd.read_csv('data.csv')

In [155]:
read

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]

In [159]:
Table_df = pd.read_csv('data.csv')


In [160]:
Table_df

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]

# always set the index=False

In [162]:
Table_df.to_csv('My_output', index=False)

In [163]:
pd.read_csv('My_output')

Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]

# now without including the index=False. this will create a column of unnamed with empty values

In [164]:
Table_df.to_csv('My_output')

In [165]:
pd.read_csv('My_output')

Unnamed: 0  Duration  Pulse  Maxpulse  Calories
0             0        60    110       130     409.1
1             1        60    117       145     479.0
2             2        60    103       135     340.0
3             3        45    109       175     282.4
4             4        45    117       148     406.0
..          ...       ...    ...       ...       ...
164         164        60    105       140     290.8
165         165        60    110       145     300.0
166         166        60    115       145     310.2
167         167        75    120       150     320.4
168         168        75    125       150     330.4

[169 rows x 5 columns]

# Reading and writting from excel files: 
* Note that Pandas can only import the data, not other attributes like images, formulars or macros etc.
* using that of the read_excel method might crash Pandas. be careful.
* might get an error asking you to intall the xlrd, if you don't have this installed, you can do that in your terminal. using pip or conda.

#  reading an excel file: 
*  Pandas excel workbook which has  of sheets and each sheet is DataFrame()

In [167]:
# pd.read_excel('work.xlsx', sheettname='sheet1')

# you can as well write into excel file

In [169]:
# df.toexcel('name_to_write.xlsx', sheet_name = 'NewSheet')

# working with HTML
* make sure you have the necessary libraries installed.


In [176]:
html_data = pd.read_html('https://www.fdic.gov/resources/resolutions/bank-failures/failed-bank-list/')

In [177]:
type(html_data)

list

In [175]:
html_data[0].head()

Bank NameBank           CityCity StateSt  CertCert  \
0                 Almena State Bank             Almena      KS     15426   
1        First City Bank of Florida  Fort Walton Beach      FL     16748   
2              The First State Bank      Barboursville      WV     14361   
3                Ericson State Bank            Ericson      NE     18265   
4  City National Bank of New Jersey             Newark      NJ     21111   

      Acquiring InstitutionAI Closing DateClosing  FundFund  
0                 Equity Bank    October 23, 2020     10538  
1   United Fidelity Bank, fsb    October 16, 2020     10537  
2              MVB Bank, Inc.       April 3, 2020     10536  
3  Farmers and Merchants Bank   February 14, 2020     10535  
4             Industrial Bank    November 1, 2019     10534

# working with SQl

# actually Pandas isn't the best python liabrary to read a SQL data. you can brows more about that, let's just do it as much as we can

In [178]:
from sqlalchemy import create_engine

In [179]:
engine = create_engine('sqlite:///:memory:')

In [181]:
new_df

A    B  C  D
0  foo  one  X  1
1  foo  one  Y  3
2  foo  two  X  2
3  bar  two  Y  5
4  bar  one  X  4
5  bar  one  Y  1

In [182]:
new_df.to_sql('my_table', engine)

In [183]:
sqldf = pd.read_sql('my_table', con=engine)

In [184]:
sqldf

index    A    B  C  D
0      0  foo  one  X  1
1      1  foo  one  Y  3
2      2  foo  two  X  2
3      3  bar  two  Y  5
4      4  bar  one  X  4
5      5  bar  one  Y  1

* donr diaturb yourself much over other documentations, as long as this lecture.
* we'll be working with CSV



In [ ]:
my_arr = [1000000000000000000000]


In [ ]:
# checking time it take for execution of a python and Numpy arrays.
# %%time is very important function on jupyter
arr1 = list(range(1000000))
arr2 = list(range(1000000, 2000000))

# numpy arrays
arr1_np = np.array(arr1)
arr2_np = np.array(arr2)

In [ ]:
%%time
result = 0
for x1, x2 in zip(arr1, arr2):
    result += x1*x2
result

In [ ]:
%%time
np.dot(arr1_np, arr2_np)

In [ ]:
# the dot function( )
help(np.dot)

In [ ]:
help(np.genfromtxt)